### PR TITLE
[FSDP] Fix `full_optim_state_dict()` hang

### LINKED ISSUE #80581: [FSDP] Saving full_state_dict for optimizer can result in hang for medium+ size models
### 🐛 Describe the bug

Using FSDP with latest nightlies (ala 627), when attempting to save the optimizer via full_state_dict call, one can readily hit an internal OOM which is shown to the user as a hang.  Example, a 500M model on V100 4 gpu single node. 
simplified code:
~~~
optim_state = FSDP.full_optim_state_dict(model, optimizer)
    if rank==0:
        torch.save(optim_state, opt_save_full_path)
~~~
result = hang with the following debug info:
hang2_optim_state

The core communication calls all complete on ranks 1-3, but rank0 hangs. 
This works fine if running smaller models ala 8M or 60M.

[edited - removed references to OOM now that we know it's not memory but the movement to cpu]

### Versions

1.13.0.dev20220627+cu113

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @ezyang

### PR DESCRIPTION
**Pull request by awgu**
Fixes https://github.com/pytorch/pytorch/issues/80581.

Context:
https://github.com/pytorch/pytorch/blob/1f08c1d3d61d1baa43f7862c3c4489487c8635d3/torch/distributed/fsdp/_optim_utils.py#L152-L163

To-Do:
I do not understand why inserting this `torch.cuda.synchronize()` prevents the `.cpu()` call from hanging and why in particular, this `torch.cuda.synchronize()` must be called on **all ranks**. If it is only called on the saving ranks (i.e. rank 0), then the hang persists.

### CODE REVIEW

File: torch/distributed/fsdp/_optim_utils.py
Comment (rohan-varma) [inline torch/distributed/fsdp/_optim_utils.py:158]: does a barrier() fix it as well?

**Review by rohan-varma — APPROVED** [review]
Stamping to unblock but we should probably try to figure out the root cause, as other use cases haven't encountered this hang AFAIK.

File: torch/distributed/fsdp/_optim_utils.py
Comment (awgu) [reply torch/distributed/fsdp/_optim_utils.py:158]: I tried with
```
dist.barrier(group=group)
```
and the hang is fixed as well.

**Comment by awgu** [comment]
After offline discussion, we will land this for now since it is not disruptive and follow-up on the explanation for why the fix works.

**Pull request closed without merging**
